flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/functions/DeclaringKeyedProcessFunction.java

```diff
@@ -38,17 +37,29 @@
  * and register further timers.
  *
  * <p><b>NOTE:</b> Access to keyed state and timers (which are also scoped to a key) is only
- * available if the {@code AsyncKeyedProcessFunction} is applied on a {@code KeyedStream}.
+ * available if the {@code DeclaringKeyedProcessFunction} is applied on a {@code KeyedStream}.
  *
  * @param <K> Type of the key.
  * @param <I> Type of the input elements.
  * @param <O> Type of the output elements.
  */
 @Internal
-public abstract class AsyncKeyedProcessFunction<K, I, O> extends AbstractAsyncStatefulRichFunction {
+public abstract class DeclaringKeyedProcessFunction<K, I, O> extends KeyedProcessFunction<K, I, O> {
 
     private static final long serialVersionUID = 1L;
 
+    /** Override and finalize this method. Please use {@link #declareProcess} instead. */
+    @Override
+    public final void processElement(I value, Context ctx, Collector<O> out) throws Exception {}
+
+    /** Override and finalize this method. Please use {@link #declareOnTimer} instead. */
+    @Override
+    public final void onTimer(long timestamp, OnTimerContext ctx, Collector<O> out)
+            throws Exception {}
+
+    /** Declaring variables before all processing. */
+    public void declareVariables(DeclarationContext context) {}
```

Comment:
Is `declareVariables` used for inheritance?

Reply:
Yes. It will be used if users want to declare some common variables for timer and record processing.